File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/application/mappers/infra/IMapper.java
package com.oddly.ddd.application.mappers.infra;

/**
 * Base interface for all mappers.
 * Mappers handle transformations between different object types:
 * - DTO ↔ BMO (Business Model Object)
 * - BMO ↔ WriteEntity
 * - ReadEntity → Response DTO
 * 
 * Mappers:
 * - Live in /application/mappers/
 * - Have Mapper suffix
 * - Implementations are in root (NO /impl/ subdirectory - no autowiring contracts)
 * - ALL transformations MUST use explicit mappers
 * 
 * @param <TRequest> The request DTO type
 * @param <TResponse> The response DTO type
 * @param <TModel> The business model object type (BMO)
 * @param <TWriteEntity> The write entity type
 * @param <TReadEntity> The read entity type
 */
public interface IMapper<TRequest, TResponse, TModel, TWriteEntity, TReadEntity> {
    /**
     * Converts a request DTO to a business model.
     * Used when receiving data from the API layer.
     * 
     * @param p_request The request DTO
     * @return The business model
     */
    TModel toModelFromRequest(TRequest p_request);

    /**
     * Converts a business model to a write entity.
     * Used by command repositories when persisting data.
     * 
     * @param p_model The business model
     * @return The write entity
     */
    TWriteEntity toWriteEntity(TModel p_model);

    /**
     * Converts a write entity to a business model.
     * Used by command repositories when reading data for commands.
     * 
     * @param p_entity The write entity
     * @return The business model
     */
    TModel toModelFromWriteEntity(TWriteEntity p_entity);

    /**
     * Converts a read entity to a response DTO.
     * Used by query repositories when returning data to the API layer.
     * Read entities are already optimized for queries, so no BMO conversion.
     * 
     * @param p_entity The read entity
     * @return The response DTO
     */
    TResponse toResponseFromReadEntity(TReadEntity p_entity);

    /**
     * Converts a business model to a response DTO.
     * Used when returning data from command operations.
     * 
     * @param p_model The business model
     * @return The response DTO
     */
    TResponse toResponseFromModel(TModel p_model);
}

File: infrastructures/rest-ddd/typescript/src/application/mappers/IMapper.ts
/**
 * Base Mapper Interface
 * ALL transformations MUST use explicit mappers.
 */

export interface IMapper<TDto, TModel, TWriteEntity, TReadEntity, TResponse> {
  /**
   * Interface for mapping between layers.
   * 
   * Rules:
   * - MANDATORY for all transformations
   * - DTO ↔ BMO ↔ Entity conversions
   * - Located in /application/mappers/
   */

  /**
   * Map Request DTO → BMO (Business Model Object).
   * Used by controllers when receiving requests.
   */
  toModelFromRequest(p_dto: TDto): TModel;

  /**
   * Map BMO → WriteEntity.
   * Used by command repositories internally.
   */
  toWriteEntity(p_model: TModel): TWriteEntity;

  /**
   * Map WriteEntity → BMO.
   * Used by command repositories internally.
   */
  toModelFromWriteEntity(p_entity: TWriteEntity): TModel;

  /**
   * Map ReadEntity → Response DTO.
   * Used by query operations.
   */
  toResponseFromReadEntity(p_entity: TReadEntity): TResponse;

  /**
   * Map BMO → Response DTO.
   * Used after command operations.
   */
  toResponseFromModel(p_model: TModel): TResponse;
}
